### PR TITLE
Add check AnsibleJob Workflow support

### PR DESCRIPTION
**Pull request by mikeshng**
Signed-off-by: Mike Ng <ming@redhat.com>

For https://github.com/stolostron/backlog/issues/26872

### CODE REVIEW

**Comment by mikeshng** [comment]
/assign @jnpacker 
/assign @xiangjingli 
/assign @elgnay 

/hold

PM wants the API layer to check if the AnsibleJob has the workflow support. Please take a look and see if this approach is correct.

Holding for backplane-operator changes.

File: controllers/clustercurator_controller.go [hunk 1]
```diff
@@ -44,6 +57,35 @@ func (r *ClusterCuratorReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		return ctrl.Result{}, nil
 	}
 
+	// If the ClusterCurator contains at least one Workflow hook type then
+	// check the AnsibleJob CRD to make sure it can support Workflow invocation.
+	if containsWorkflowType(curator) {
+		err := r.checkWorkflowSupport(ctx, curator)
+		if err != nil {
```
Comment (elgnay) [inline controllers/clustercurator_controller.go:64]: It seems errors returned by r.checkWorkflowSupport() may have various reasons. WorkflowNotSupported is one of them.

File: controllers/clustercurator_controller.go [hunk 2]
```diff
@@ -44,6 +57,35 @@ func (r *ClusterCuratorReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
+			meta.SetStatusCondition(&curator.Status.Conditions, metav1.Condition{
+				Type:    statusTypeInitiated,
+				Status:  metav1.ConditionFalse,
+				Reason:  statusReasonWorkflowNotSupported,
+				Message: "The AnsibleJob CRD does not support Workflow invocation"})
```
[the remark this answers is not in the document]
Comment (mikeshng) [reply controllers/clustercurator_controller.go:69]: Hi @showeimer @KevinFCormier 

This PR is for the backend API check as discussed on Slack. Do you think there are cycles for the UI team to handle an additional status condition? 
The new status condition works like this:
- Only populated if the ClusterCurator contains a Workflow type in one of the hooks.
- If the AnsibleJob CRD doesn't support workflow invocation then there is a status condition type `Initiated` which is set to false with the reason `WorkflowNotSupported`
- If the AnsibleJob CRD does support workflow then the status condition `Initiated` is set to true and the reason is `WorkflowSupported`
- Based on Le Yang's feedback there might be another reason when `Initiated` is false with the reason `FailedWorkflowSupportCheck`

[the remark this answers is not in the document]
Comment (mikeshng) [reply controllers/clustercurator_controller.go:69]: @elgnay it looks like the UI squad won't have the cycle to consume this new condition. So perhaps for this iteration, we should not introduce a new status condition. We will log that the Workflow invocation is not supported and just treat it as no work needs to be done. 
There is already a similar scenario in the existing code so it won't be outside of the norm. See: https://github.com/stolostron/cluster-curator-controller/blob/main/controllers/clustercurator_controller.go#L60-L64

[the remark this answers is not in the document]
Comment (KevinFCormier) [reply controllers/clustercurator_controller.go:69]: @mikeshng @elgnay Sorry, I missed the earlier comment about this and assumed that we'd still get some sort of failure condition. I think we would prioritize this above adding a warning.

I will try applying this patch and seeing if I can contain the necessary changes.

**Comment by mikeshng** [comment]
This API layer check was more work and requires cohesion from multiple teams than we originally anticipated. As agreed to, closing this PR for now. It will be reopen/work on in the next release.

**Pull request closed without merging**